client/network/sync/src/engine.rs

```diff
@@ -616,42 +615,32 @@ where
 					continue
 				}
 
-				// reset the peer activity timers so they don't expire right away after
-				// the initial wait is done.
-				for info in self.peers.values_mut() {
-					info.last_notification_received = Instant::now();
-					info.last_notification_sent = Instant::now();
-				}
-
 				self.syncing_started = None;
+				self.last_notification_sent = Instant::now();
+				self.last_notification_received = Instant::now();
 			}
 
-			// go over all connected peers and check if any of them have been idle for a while. Idle
-			// in this case means that we haven't sent or received block announcements to/from this
-			// peer. If that is the case, because of #5685, it could be that the block announces
-			// substream is not actually open and and this peer is just wasting a slot and is should
-			// be replaced with some other node that is willing to send us block announcements.
-			for (id, peer) in self.peers.iter() {
-				// because of a delay between disconnecting a peer in `SyncingEngine` and getting
-				// the response back from `Protocol`, a peer might be reported and disconnect
-				// multiple times. To prevent this from happening (until the underlying issue is
-				// fixed), keep track of evicted peers and report and disconnect them only once.
-				if self.evicted.contains(id) {
-					continue
-				}
+			// if syncing hasn't sent or received any blocks within `INACTIVITY_EVICT_THRESHOLD`,
+			// it means the local node has stalled and is connected to peers who either don't
+			// consider it connected or are also all stalled. In order to unstall the node,
+			// disconnect all peers and allow `ProtocolController` to establish new connections.
+			let last_received_late =
+				self.last_notification_received.elapsed() > INACTIVITY_EVICT_THRESHOLD;
+			let last_sent_late = self.last_notification_sent.elapsed() > INACTIVITY_EVICT_THRESHOLD;
 
-				let last_received_late =
-					peer.last_notification_received.elapsed() > INACTIVITY_EVICT_THRESHOLD;
-				let last_sent_late =
-					peer.last_notification_sent.elapsed() > INACTIVITY_EVICT_THRESHOLD;
+			if last_received_late && last_sent_late {
```

Comment:
> if syncing hasn't sent **or** received any blocks`

There should be `||` instead of `&&`. And because of `||` there is no need to use two variables last_notification_received and last_notification_sent, its enough to use only one - something like last_notification_io.

Reply:
Fair point. There was lengthy discussion on this in the last PR but I agree that if we track the state not on peer-level but "local node level", `last_notification_io` sounds sensible.